### PR TITLE
feat: add goroutine watcher

### PR DESCRIPTION
**Pull request by jake-shin0**
- `queryer` was managed as a separate dir and its implementation was divided. (cgroup & runtime)
- `Queryer` interface method to public.
- As I was dividing `queryer`, I also wanted to manage `profiler` in dir, but I did not consider this part.
- Since both monitor and profile are managed through the profile singleton object in pprof lib, I considered implementing it by injecting and managing a common singleton object in `queryer` and `profiler`, but did not do so.
  - The singleton object inside the pprof lib is managed as a map, making its usability convenient, but the public function it returns is a list, making its usability a bit ambiguous.
  - if creating an internal singleton object, it must be taken care of concurrency issues.
  - It is not a frequently called object, and its call frequency is determined by `ticker`, so pprof lib is used in both `queryer` and `profiler`.

### CODE REVIEW

File: autopprof.go [hunk 1]
```diff
@@ -53,8 +64,9 @@ type autoPprof struct {
 	reportBoth bool
```
Comment (jake-shin0) [inline autopprof.go:64]: 해당 필드는 cpu 와 memory, 단 두개만 모니터링할때만 유효한 옵션으로 runtime moniter 가 늘어날수록 그 의미가 없어지는데, 별도의 버전에서 제거하는 것은 어떨지 @mingrammer 의견 부탁해요-!

File: autopprof_test.go [hunk 2]
```diff
@@ -858,13 +1020,13 @@ func BenchmarkLightJob(b *testing.B) {
 
 func BenchmarkLightJobWithWatchCPUUsage(b *testing.B) {
```
Comment (dlsrb6342) [inline autopprof_test.go:1021]: i think we need to benckmark goroutine watcher too.

File: go.sum [hunk 3]
```diff
@@ -93,4 +93,4 @@ golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1/go.mod h1:I/5z698sn9Ka8T
 gopkg.in/check.v1 v0.0.0-20161208181325-20d25e280405/go.mod h1:Co6ibVJAznAaIkqp8huTwlJQCZ016jof/cbN4VW5Yz0=
 gopkg.in/check.v1 v1.0.0-20180628173108-788fd7840127/go.mod h1:Co6ibVJAznAaIkqp8huTwlJQCZ016jof/cbN4VW5Yz0=
 gopkg.in/yaml.v3 v3.0.0-20200313102051-9f266ea9e77c h1:dUUwHk2QECo/6vqA44rthZ8ie2QXMNeKRTHCNY2nXvo=
-gopkg.in/yaml.v3 v3.0.0-20200313102051-9f266ea9e77c/go.mod h1:K4uyk7z7BCEPqu6E+C64Yfv1cQ7kz7rIZviUmN+EgEM=
+gopkg.in/yaml.v3 v3.0.0-20200313102051-9f266ea9e77c/go.mod h1:K4uyk7z7BCEPqu6E+C64Yfv1cQ7kz7rIZviUmN+EgEM=
```
Comment (dlsrb6342) [inline go.sum:96]: nit, add new line or run `go mod tidy`

File: autopprof.go [hunk 4]
```diff
@@ -65,7 +76,12 @@ var globalAp *autoPprof
 
 // Start configures and runs the autopprof process.
 func Start(opt Option) error {
-	qryer, err := newQueryer()
+	cgroupQryer, err := queryer.NewCgroupQueryer()
+	if err != nil {
+		return err
+	}
+
+	runtimeQryer, err := queryer.NewRuntimeQueryer()
```
Comment (suripark) [inline autopprof.go:84]: typo?

File: report/slack.go [hunk 5]
```diff
@@ -85,3 +86,25 @@ func (s *SlackReporter) ReportHeapProfile(
 	}
 	return nil
 }
+
+// ReportGoroutineProfile sends the goroutine profiling data to the Slack.
+func (s *SlackReporter) ReportGoroutineProfile(
+	ctx context.Context, r io.Reader, gi GoroutineInfo,
+) error {
+	hostname, _ := os.Hostname() // Don't care about this error.
```
Comment (suripark) [inline report/slack.go:94]: p999; I think this comment is vague, how about just removing it?

File: queryer/cgroupv1.go [hunk 6]
```diff
@@ -45,7 +45,38 @@ func newCgroupsV1() *cgroupV1 {
...
+}
+
+func (c *cgroupV1) MemUsage() (float64, error) {
+	stat, err := c.stat()
+	if err != nil {
+		return 0, err
+	}
+	var (
+		sm    = stat.Memory
+		usage = sm.Usage.Usage - sm.InactiveFile
```
Comment (suripark) [inline queryer/cgroupv1.go:73]: nice

File: queryer/cgroupv2.go [hunk 7]
```diff
@@ -49,7 +49,38 @@ func newCgroupsV2() *cgroupV2 {
...
+		return 0, err
+	}
+	c.snapshotCPUUsage(stat.CPU.UsageUsec) // In microseconds.
+
+	// Calculate the usage only if there are enough snapshots.
+	if !c.q.isFull() {
+		return 0, nil
+	}
+
+	s1, s2 := c.q.head(), c.q.tail()
```
Comment (suripark) [inline queryer/cgroupv2.go:64]: circular queue ! nice

**Review by suripark — APPROVED** [review]

File: autopprof_test.go (hunk 2 above)
Comment (jake-shin0) [reply autopprof_test.go:1023]: I artificially created a lot of goroutines and ran a benchmark test. This is my first time doing a test like this. What should I fix?
```
BenchmarkLightAsyncJob-5                                  376760             32056 ns/op            7040 B/op        352 allocs/op
BenchmarkLightAsyncJobWithWatchGoroutineCount-5           350994             33689 ns/op            7040 B/op        351 allocs/op
BenchmarkHeavyAsyncJob-5                                     414          29095456 ns/op         6001969 B/op     300096 allocs/op
BenchmarkHeavyAsyncJobWithWatchGoroutineCount-5              403          29474789 ns/op         5972177 B/op     298607 allocs/op
```

File: autopprof.go (hunk 1 above)
Comment (mingrammer) [reply autopprof.go:63]: We should introduce a new option, reportAll, which reports all profile results when any threshold is exceeded. And deprecate the reportBoth.

File: autopprof.go [hunk 8]
```diff
@@ -291,6 +313,68 @@ func (ap *autoPprof) reportHeapProfile(memUsage float64) error {
...
+func (ap *autoPprof) reportGoroutineProfile(goroutineCount int) error {
+	b, err := ap.profiler.profileGoroutine()
+	if err != nil {
+		return fmt.Errorf("autopprof: failed to profile the goroutine: %w", err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), reportTimeout)
+	defer cancel()
+
+	ci := report.GoroutineInfo{
```
Comment (mingrammer) [inline autopprof.go:367]:
```suggestion
	gi := report.GoroutineInfo{
```

File: queryer/cgroupv1.go [hunk 9]
```diff
@@ -30,7 +30,7 @@ type cgroupV1 struct {
 
 	cpuQuota float64
 
-	q cpuUsageSnapshotQueuer
+	q CpuUsageSnapshotQueuer
```
Comment (mingrammer) [inline queryer/cgroupv1.go:33]: Please use "CPU" in capital case.

Comment (mingrammer) [inline queryer/cgroupv1.go:33]:
```suggestion
	q CPUUsageSnapshotQueuer
```

File: queryer/cgroupv1.go [hunk 10]
```diff
@@ -45,7 +45,38 @@ func newCgroupsV1() *cgroupV1 {
 	}
 }
 
-func (c *cgroupV1) setCPUQuota() error {
+func (c *cgroupV1) CpuUsage() (float64, error) {
```
Comment (mingrammer) [inline queryer/cgroupv1.go:48]:
```suggestion
func (c *cgroupV1) CPUUsage() (float64, error) {
```

File: queryer/cgroupv2.go [hunk 11]
```diff
@@ -49,7 +49,38 @@ func newCgroupsV2() *cgroupV2 {
 	}
 }
 
-func (c *cgroupV2) setCPUQuota() error {
+func (c *cgroupV2) CpuUsage() (float64, error) {
```
Comment (mingrammer) [inline queryer/cgroupv2.go:52]:
```suggestion
func (c *cgroupV2) CPUUsage() (float64, error) {
```

File: queryer/cgroupv2.go [hunk 12]
```diff
@@ -34,7 +34,7 @@ type cgroupV2 struct {
 
 	cpuQuota float64
 
-	q cpuUsageSnapshotQueuer
+	q CpuUsageSnapshotQueuer
```
Comment (mingrammer) [inline queryer/cgroupv2.go:37]:
```suggestion
	q CPUUsageSnapshotQueuer
```

File: queryer/cgroupv1_test.go [hunk 13]
```diff
@@ -10,7 +10,7 @@ import (
 	"github.com/containerd/cgroups"
 )
 
-func TestCgroupV1_cpuUsage(t *testing.T) {
+func TestCgroupV1_CpuUsage(t *testing.T) {
```
Comment (mingrammer) [inline queryer/cgroupv1_test.go:13]:
```suggestion
func TestCgroupV1_CPUUsage(t *testing.T) {
```

File: queryer/cgroupv2_test.go [hunk 14]
```diff
@@ -10,7 +10,7 @@ import (
 	"github.com/containerd/cgroups"
 )
 
-func TestCgroupV2_cpuUsage(t *testing.T) {
+func TestCgroupV2_CpuUsage(t *testing.T) {
```
Comment (mingrammer) [inline queryer/cgroupv2_test.go:13]:
```suggestion
func TestCgroupV2_CPUUsage(t *testing.T) {
```

File: queryer/queue.go [hunk 15]
```diff
@@ -1,10 +1,10 @@
-package autopprof
+package queryer
 
 import "time"
 
-// cpuUsageSnapshotQueue is a circular queue of cpuUsageSnapshot.
+// CpuUsageSnapshotQueue is a circular queue of cpuUsageSnapshot.
 // It doesn't implement dequeue() method because it's not needed.
-type cpuUsageSnapshotQueuer interface {
+type CpuUsageSnapshotQueuer interface {
```
Comment (mingrammer) [inline queryer/queue.go:7]: I think we don't have to make it public.

File: queryer/runtime_test.go [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+	}
+
+	addedGoroutineCnt := r.GoroutineCount()
+	if addedGoroutineCnt != initGoroutineCnt+goroutineCnt {
+		t.Errorf("GoroutineCount() = %d; want is %d", addedGoroutineCnt, initGoroutineCnt+1)
+	}
+
+	wg.Wait()
+
+	endGoroutineCnt := r.GoroutineCount()
```
Comment (mingrammer) [inline queryer/runtime_test.go:35]: remainedGoroutineCount?

File: queryer/runtime_test.go [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+	goroutineCnt := 1000
+	for i := 0; i < goroutineCnt; i++ {
+		wg.Add(1)
+		go func() {
+			time.Sleep(500 * time.Millisecond)
+			wg.Done()
+		}()
+	}
+
+	addedGoroutineCnt := r.GoroutineCount()
```
Comment (mingrammer) [inline queryer/runtime_test.go:28]: processingGoroutineCnt?

File: autopprof.go (hunk 1 above)
Comment (jake-shin0) [reply autopprof.go:63]: Good idea! I will create this idea as a separate PR.

File: autopprof.go (hunk 4 above)
Comment (jake-shin0) [reply autopprof.go:84]: No. this is normal naming of this repository

**Comment by jake-shin0** [comment]
@mingrammer 
macOS 11 image on runner was deprecated

**Review by mingrammer — APPROVED** [review]
LGTM!

**Pull request merged**